cmd/watch/main.go

```diff
@@ -0,0 +1,221 @@
+package main
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/deref/exo/util/cmdutil"
+	"github.com/fsnotify/fsnotify"
+)
+
+type runState int
+
+const (
+	runStateInitial runState = iota
+	runStateRunning
+	runStateRestarting
+	runStateWaitingChildExit
+)
+
+func (s runState) String() string {
+	switch s {
+	case runStateInitial:
+		return "Initial"
+	case runStateRunning:
+		return "Running"
+	case runStateRestarting:
+		return "Restarting"
+	case runStateWaitingChildExit:
+		return "WaitingChildExit"
+	}
+	return "<invalid state>"
+}
+
+func main() {
+	cmd, err := cmdutil.ParseArgs(os.Args)
+	if err != nil {
+		cmdutil.Fatalf("parsing arguments: %w", err)
+	}
+
+	if len(cmd.Args) == 0 {
+		cmdutil.Fatalf("expected command to execute")
+	}
+
+	reload := make(chan struct{}, 1)
+	childStarted := make(chan struct{}, 1)
+	childStopped := make(chan struct{}, 1)
+	stopSignals := make(chan os.Signal)
+	done := make(chan struct{})
+
+	var child *exec.Cmd
+	startNewChild := func() {
+		var err error
+		if child, err = startCommand(cmd.Args, childStarted, childStopped); err != nil {
+			cmdutil.Fatalf("starting child process: %w", err)
+		}
+	}
+
+	signal.Notify(stopSignals, os.Interrupt, syscall.SIGTERM, syscall.SIGQUIT)
+
+	// Main loop
+	go func() {
+		state := runStateInitial
+		for {
+			select {
+			case sig := <-stopSignals:
+				gracefullyShutdown(child, sig)
+				state = runStateWaitingChildExit
+
+			default:
+				// Did not receive a stop signal since last iteration - continue.
+			}
+
+			select {
+			case _ = <-reload:
+				// Got additional reload notifications - ignore.
+			default:
+				// No additional notifications.
+			}
+
+			switch state {
+			case runStateInitial:
+				<-childStarted
+				state = runStateRunning
+
+			case runStateRunning:
+				select {
+				case _ = <-reload:
+					fmt.Println("Files changed. Restarting process.")
+					gracefullyShutdown(child, syscall.SIGTERM)
+					state = runStateRestarting
+
+				case sig := <-stopSignals:
+					gracefullyShutdown(child, sig)
+					state = runStateWaitingChildExit
+
+				case _ = <-childStopped:
+					// Child stopped when we were not expecting a restart, so exit.
+					done <- struct{}{}
+				}
+
+			case runStateRestarting:
+				gracefulShutdownTimer := time.NewTimer(time.Second * time.Duration(10))
```

Comment:
since this timer is only used locally here, not looped over or anything like that, I think it is safe to simply do `case <-time.After(...)`, as that will produce a one-element buffered channel and when the timer fires, the sending code will terminate and the channel will get GC'd, no need to call `Stop`. Tho I guess this does cleanup the timer sooner, so maybe it's worth being verbose?